src/Servers/Kestrel/Core/src/Internal/Http/Http1OutputProducer.cs

```diff
@@ -188,5 +277,56 @@ public Task Write100ContinueAsync()
                     cancellationToken);
             }
         }
+
+        private Memory<byte> GetChunkedMemory(int sizeHint)
+        {
+            // The max size of a chunk will be the size of memory returned from the PipeWriter (today 4096)
+            // minus 5 for the max chunked prefix size and minus 2 for the chunked ending, leaving a total of
+            // 4089.
+
+            if (_currentChunkMemory.Length == 0)
+            {
+                // First time calling GetMemory
+                _currentChunkMemory = _pipeWriter.GetMemory(sizeHint);
+            }
+
+            var memoryMaxLength = _currentChunkMemory.Length - BeginChunkLengthMax - EndChunkLength;
+            if (_advancedBytesForChunk == memoryMaxLength)
+            {
+                // Chunk is completely written, commit it to the pipe so GetMemory will return a new chunk of memory.
+                CommitChunkToPipe();
```

Comment:
Does this mean we can implicitly commit what should be uncommitted bytes in PipeWriter.GetMemory/Span()?

Reply:
@pakrym

That's not how Pipe's normally work, right? I just double checked the source, and Advance only advances the write head not the read head. You have to flush or complete the PipeWriter to commit data.

Code that writes to the buffer in multiple segments and then goes back to change something in a previous segment (just like we were considering doing with the transport pipe to efficiently chunk responses) would break.

Reply:
> Code that writes to the buffer in multiple segments and then goes back to change something in a previous segment (just like we were considering doing with the transport pipe to efficiently chunk responses) would break.

Didn't we decide that this is an unsupported scenario?

Reply:
That behavior could be pretty useful. If at all possible, I'd like to support it.

Reply:
Yes! Come to the dark side. We need to support this 😄

Reply:
I don't understand, we were in the same meeting where we decided that's not a good idea without giving a consumer of IBufferWriter a way to notify other components that memory is preserved.


Even with pipe implementation holding onto Memory is safe only until flush is called which is often not controlled by IBufferWriter consumer. Like in the following sample:

``` C#
var writer = new SomeLenghtPrefixSerializer(response.BodyPipe);
foreach (var o in objects)
{
   writer.WriteObject(o);
   // Is this safe to do or is writer still holding onto memory?
   await response.BodyPipe.FlushAsync();
}
```

Getting this kind of assumption wrong can lead to severe consequences - considering that we reuse blocks response intended for one client can be sent to another one.

Another reason is that IBufferWriter implementation can be transforming bytes synchronously (encryption/compression) after they were advanced.